src/target/riscv/riscv_reg.c

```diff
@@ -795,97 +782,17 @@ int riscv_reg_flush_all(struct target *target)
- * If `write_through` is true, it is ensured that the value of the target's
- * register is set to be equal to the `value` argument. The cached value is
- * updated if the register is cacheable.
- * TODO: Currently `reg->get/set` is implemented in terms of
- * `riscv_get/set_register`.  However, the intention behind
- * `riscv_get/set_register` is to work with the cache, therefore it accesses
- * and modifyes register cache directly.  The idea is to implement
- * `riscv_get/set_register` in terms of `riscv_reg_impl_cache_entry` and
- * `reg->get/set`.
- */
-static int riscv_set_or_write_register(struct target *target,
-		enum gdb_regno regid, riscv_reg_t value, bool write_through)
-{
-	RISCV_INFO(r);
-	assert(r);
-	if (r->dtm_version == DTM_DTMCS_VERSION_0_11)
-		return riscv011_set_register(target, regid, value);
-
-	keep_alive();
-
-	if (regid == GDB_REGNO_PC) {
-		return riscv_set_or_write_register(target, GDB_REGNO_DPC, value, write_through);
-	} else if (regid == GDB_REGNO_PRIV) {
-		riscv_reg_t dcsr;
-
-		if (riscv_reg_get(target, &dcsr, GDB_REGNO_DCSR) != ERROR_OK)
-			return ERROR_FAIL;
-		dcsr = set_field(dcsr, CSR_DCSR_PRV, get_field(value, VIRT_PRIV_PRV));
-		dcsr = set_field(dcsr, CSR_DCSR_V, get_field(value, VIRT_PRIV_V));
-		return riscv_set_or_write_register(target, GDB_REGNO_DCSR, dcsr, write_through);
-	}
-
-	struct reg *reg = riscv_reg_impl_cache_entry(target, regid);
-	assert(riscv_reg_impl_is_initialized(reg));
-
-	if (!reg->exist) {
-		LOG_TARGET_DEBUG(target, "Register %s does not exist.", reg->name);
-		return ERROR_FAIL;
-	}
-
-	if (target->state != TARGET_HALTED) {
-		LOG_TARGET_DEBUG(target,
-				"Target not halted, writing to target: %s <- 0x%" PRIx64,
-				reg->name, value);
-		return riscv013_set_register(target, regid, value);
-	}
-
-	const bool need_to_write = !reg->valid || reg->dirty ||
-		value != buf_get_u64(reg->value, 0, reg->size);
-	const bool cacheable = riscv_reg_impl_gdb_regno_cacheable(regid, need_to_write);
-
-	if (!cacheable || (write_through && need_to_write)) {
-		LOG_TARGET_DEBUG(target,
-				"Writing to target: %s <- 0x%" PRIx64 " (cacheable=%s, valid=%s, dirty=%s)",
-				reg->name, value, cacheable ? "true" : "false",
-				reg->valid ? "true" : "false",
-				reg->dirty ? "true" : "false");
-		if (riscv013_set_register(target, regid, value) != ERROR_OK)
-			return ERROR_FAIL;
-
-		reg->dirty = false;
-	} else {
-		reg->dirty = need_to_write;
+		if (reg->type->flush(reg) != ERROR_OK)
+			res = ERROR_FAIL;
 	}
-
-	buf_set_u64(reg->value, 0, reg->size, value);
-	reg->valid = cacheable;
-
-	LOG_TARGET_DEBUG(target,
-			"Wrote 0x%" PRIx64 " to %s (cacheable=%s, valid=%s, dirty=%s)",
-			value, reg->name, cacheable ? "true" : "false",
-			reg->valid ? "true" : "false",
-			reg->dirty ? "true" : "false");
-	return ERROR_OK;
+	if (res == ERROR_OK)
+		LOG_TARGET_DEBUG(target, "Flush of register cache completed successfully.");
+	else
+		LOG_TARGET_DEBUG(target, "Flush of register cache failed.");
```

Comment:
1) Should this be an error instead of debug print?

2) It turns out that the OpenOCD upstream code style does not like punctuation at the end of log messages. I have created an easy issue to fix all the messages in `src/target/riscv`: https://github.com/riscv-collab/riscv-openocd/issues/1212

```suggestion
		LOG_TARGET_ERROR(target, "Flush of register cache failed");
```

Reply:
(1) Addressed. Thanks for the catch!
(2) Looked through the changes -- should be fixed.